File: src/dispatch/providers/anthropic.py
"""Anthropic provider — calls Claude API for audit/fix/evaluate/write tasks.

HTTP-level timeout on every call. Schema validation via validate_and_retry.
"""

import json
import os
from typing import Optional

import anthropic

from ..llm import LLMDispatcher
from ..schema import AuditResult, EvaluationResult, FixResult, WriteResult
from ..validation import validate_and_retry

DEFAULT_MODEL = "claude-sonnet-4-20250514"
DEFAULT_TIMEOUT = 120.0  # seconds
DEFAULT_MAX_TOKENS = 4096


class AnthropicProvider(LLMDispatcher):
    """Anthropic Claude provider with HTTP-level timeouts and schema validation."""

    def __init__(
        self,
        api_key: Optional[str] = None,
        model: str = DEFAULT_MODEL,
        timeout: float = DEFAULT_TIMEOUT,
        max_tokens: int = DEFAULT_MAX_TOKENS,
        max_retries: int = 2,
    ):
        key = api_key or os.environ.get("ANTHROPIC_API_KEY", "")
        if not key:
            raise ValueError(
                "ANTHROPIC_API_KEY not set. Provide api_key or set the environment variable."
            )
        self.client = anthropic.Anthropic(
            api_key=key,
            timeout=timeout,
        )
        self.model = model
        self.max_tokens = max_tokens
        self.max_retries = max_retries

    def _call(
        self,
        system: str,
        user: str,
        error_feedback: Optional[str] = None,
    ) -> str:
        """Make a single API call and return the text response."""
        messages = [{"role": "user", "content": user}]
        if error_feedback:
            messages = [
                {"role": "user", "content": user},
                {
                    "role": "assistant",
                    "content": "I'll fix the formatting issues.",
                },
                {
                    "role": "user",
                    "content": f"Your previous response had a schema error: {error_feedback}\n\nPlease try again with valid JSON.",
                },
            ]

        response = self.client.messages.create(
            model=self.model,
            max_tokens=self.max_tokens,
            system=system,
            messages=messages,
        )
        # Extract text from response
        text = ""
        for block in response.content:
            if block.type == "text":
                text += block.text
        return _extract_json(text)

    def audit(
        self,
        files: list[str],
        dimensions: list[str],
        skill_context: str,
    ) -> AuditResult:
        system = _audit_system_prompt(dimensions, skill_context)
        user = _audit_user_prompt(files)

        def call(**kwargs):
            return self._call(system, user, **kwargs)

        return validate_and_retry(call, AuditResult, self.max_retries)

    def fix(
        self,
        finding_id: str,
        finding_description: str,
        file_path: str,
        file_content: str,
        skill_context: str,
    ) -> FixResult:
        system = _fix_system_prompt(skill_context)
        user = _fix_user_prompt(finding_id, finding_description, file_path, file_content)

        def call(**kwargs):
            return self._call(system, user, **kwargs)

        return validate_and_retry(call, FixResult, self.max_retries)

    def evaluate_independence(
        self,
        pass_a: AuditResult,
        pass_b: AuditResult,
    ) -> EvaluationResult:
        system = "You evaluate whether two audit passes were independent. Respond with JSON: {\"independent\": bool, \"rationale\": str}"
        user = (
            f"Pass A audited {pass_a.files_audited} on dimensions {pass_a.dimensions_checked}, "
            f"found {len(pass_a.findings)} findings.\n"
            f"Pass B audited {pass_b.files_audited} on dimensions {pass_b.dimensions_checked}, "
            f"found {len(pass_b.findings)} findings.\n\n"
            "Were these passes independent (different scope, different approach)?"
        )

        def call(**kwargs):
            return self._call(system, user, **kwargs)

        return validate_and_retry(call, EvaluationResult, self.max_retries)

    def write(self, spec: str, skill_context: str) -> WriteResult:
        system = _write_system_prompt(skill_context)
        user = f"Write the following:\n\n{spec}"

        def call(**kwargs):
            return self._call(system, user, **kwargs)

        return validate_and_retry(call, WriteResult, self.max_retries)


def _extract_json(text: str) -> str:
    """Extract JSON from LLM response text.

    Handles cases where the JSON is wrapped in markdown code blocks.
    """
    text = text.strip()
    # Try to find JSON in code blocks
    if "```json" in text:
        start = text.index("```json") + 7
        end = text.index("```", start)
        return text[start:end].strip()
    if "```" in text:
        start = text.index("```") + 3
        end = text.index("```", start)
        return text[start:end].strip()
    # Try raw JSON (starts with { or [)
    for i, c in enumerate(text):
        if c in "{[":
            # Find matching end
            depth = 0
            for j in range(i, len(text)):
                if text[j] in "{[":
                    depth += 1
                elif text[j] in "}]":
                    depth -= 1
                    if depth == 0:
                        return text[i : j + 1]
    return text


def _audit_system_prompt(dimensions: list[str], skill_context: str) -> str:
    dims = ", ".join(dimensions)
    context = f"\n\nContext:\n{skill_context}" if skill_context else ""
    return (
        f"You are a code auditor. Audit the provided files on these dimensions: {dims}.\n\n"
        "Respond with JSON matching this schema:\n"
        '{"findings": [{"id": str, "file": str, "dimension": str, '
        '"severity": "high"|"medium"|"low", "description": str, "suggested_fix": str}], '
        '"files_audited": [str], "dimensions_checked": [str]}\n\n'
        "If no issues found, return empty findings array."
        f"{context}"
    )


def _audit_user_prompt(files: list[str]) -> str:
    parts = ["Audit these files:\n"]
    for f in files:
        try:
            with open(f) as fh:
                content = fh.read()
            parts.append(f"--- {f} ---\n{content}\n")
        except (FileNotFoundError, OSError):
            parts.append(f"--- {f} --- (file not found)\n")
    return "\n".join(parts)


def _fix_system_prompt(skill_context: str) -> str:
    context = f"\n\nContext:\n{skill_context}" if skill_context else ""
    return (
        "You fix code issues. Apply the suggested fix and respond with JSON:\n"
        '{"success": bool, "files_modified": [str], "description": str}'
        f"{context}"
    )


def _fix_user_prompt(
    finding_id: str,
    finding_description: str,
    file_path: str,
    file_content: str,
) -> str:
    return (
        f"Finding {finding_id}: {finding_description}\n\n"
        f"File: {file_path}\n"
        f"Content:\n{file_content}\n\n"
        "Fix this issue."
    )


def _write_system_prompt(skill_context: str) -> str:
    context = f"\n\nContext:\n{skill_context}" if skill_context else ""
    return (
        "You write code and documentation. Respond with JSON:\n"
        '{"content": str, "files_written": [str], "description": str}'
        f"{context}"
    )

File: rust/src/dispatch/anthropic.rs
//! Anthropic API provider — makes real LLM calls via reqwest.
//!
//! Reads ANTHROPIC_API_KEY from environment. Supports model routing:
//! - fast: claude-haiku-4-5-20251001
//! - standard: claude-sonnet-4-6-20250514
//! - frontier: claude-opus-4-6-20250514

use super::{AuditFinding, AuditResult, EvaluationResult, FixResult, LLMDispatcher, Severity, WriteResult};
use serde::{Deserialize, Serialize};

const API_URL: &str = "https://api.anthropic.com/v1/messages";
const API_VERSION: &str = "2023-06-01";

/// Model tier for routing.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ModelTier {
    Fast,
    Standard,
    Frontier,
}

impl ModelTier {
    pub fn model_id(&self) -> &'static str {
        match self {
            Self::Fast => "claude-haiku-4-5-20251001",
            Self::Standard => "claude-sonnet-4-6-20250514",
            Self::Frontier => "claude-opus-4-6-20250514",
        }
    }
}

/// Anthropic LLM provider.
pub struct AnthropicProvider {
    client: reqwest::Client,
    api_key: String,
    tier: ModelTier,
    max_tokens: u32,
    call_count: u64,
}

impl AnthropicProvider {
    /// Create from environment variable ANTHROPIC_API_KEY.
    pub fn from_env(tier: ModelTier) -> anyhow::Result<Self> {
        let api_key = std::env::var("ANTHROPIC_API_KEY")
            .map_err(|_| anyhow::anyhow!("ANTHROPIC_API_KEY not set"))?;
        Ok(Self {
            client: reqwest::Client::new(),
            api_key,
            tier,
            max_tokens: 4096,
            call_count: 0,
        })
    }

    /// Create with explicit API key.
    pub fn new(api_key: &str, tier: ModelTier) -> Self {
        Self {
            client: reqwest::Client::new(),
            api_key: api_key.to_string(),
            tier,
            max_tokens: 4096,
            call_count: 0,
        }
    }

    pub fn call_count(&self) -> u64 {
        self.call_count
    }

    async fn call_api(&mut self, system: &str, user: &str) -> anyhow::Result<String> {
        self.call_count += 1;

        let body = serde_json::json!({
            "model": self.tier.model_id(),
            "max_tokens": self.max_tokens,
            "system": system,
            "messages": [{"role": "user", "content": user}],
        });

        let resp = self.client
            .post(API_URL)
            .header("x-api-key", &self.api_key)
            .header("anthropic-version", API_VERSION)
            .header("content-type", "application/json")
            .json(&body)
            .send()
            .await?;

        let status = resp.status();
        let text = resp.text().await?;

        if !status.is_success() {
            return Err(anyhow::anyhow!("Anthropic API error {}: {}", status, text));
        }

        let parsed: serde_json::Value = serde_json::from_str(&text)?;
        let content = parsed
            .get("content")
            .and_then(|c| c.as_array())
            .and_then(|arr| arr.first())
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("")
            .to_string();

        Ok(content)
    }
}

/// Schema validation: parse LLM JSON output, retry on failure.
pub fn validate_json<T: serde::de::DeserializeOwned>(raw: &str) -> anyhow::Result<T> {
    // Try to find JSON in the response (LLMs sometimes wrap in markdown)
    let json_str = if let Some(start) = raw.find('{') {
        if let Some(end) = raw.rfind('}') {
            &raw[start..=end]
        } else {
            raw
        }
    } else if let Some(start) = raw.find('[') {
        if let Some(end) = raw.rfind(']') {
            &raw[start..=end]
        } else {
            raw
        }
    } else {
        raw
    };

    serde_json::from_str(json_str).map_err(|e| anyhow::anyhow!("Schema validation failed: {e}"))
}

/// Validate and retry pattern: call LLM, validate output, retry on schema failure.
pub fn validate_and_retry<T, F>(
    mut call: F,
    max_retries: usize,
) -> anyhow::Result<T>
where
    T: serde::de::DeserializeOwned,
    F: FnMut() -> anyhow::Result<String>,
{
    for attempt in 0..=max_retries {
        let raw = call()?;
        match validate_json::<T>(&raw) {
            Ok(parsed) => return Ok(parsed),
            Err(_) if attempt < max_retries => continue,
            Err(e) => return Err(anyhow::anyhow!(
                "Schema validation failed after {} attempts: {e}",
                max_retries + 1
            )),
        }
    }
    unreachable!()
}

// Note: LLMDispatcher trait uses sync methods, but Anthropic needs async.
// For now, we provide the async call_api method and sync wrappers are
// implemented at the integration layer (tokio::runtime::Handle::current().block_on).
// The trait will be made async in a future refactor.

impl LLMDispatcher for AnthropicProvider {
    fn audit(
        &mut self,
        files: &[String],
        dimensions: &[String],
        skill_context: &str,
    ) -> anyhow::Result<AuditResult> {
        let system = format!(
            "You are a code auditor. Audit the provided files on these dimensions: {}.\n\
             Respond with JSON matching this schema: \
             {{\"findings\": [{{\"id\": \"f1\", \"file\": \"path\", \"dimension\": \"dim\", \
             \"severity\": \"high|medium|low\", \"description\": \"desc\", \"suggested_fix\": \"fix\"}}]}}\n\
             {}",
            dimensions.join(", "),
            if skill_context.is_empty() { "" } else { skill_context },
        );
        let user = format!("Audit these files:\n{}", files.join("\n"));

        // Sync wrapper — use tokio runtime
        let rt = tokio::runtime::Handle::try_current()
            .map_err(|_| anyhow::anyhow!("No tokio runtime"))?;
        let response = rt.block_on(self.call_api(&system, &user))?;

        #[derive(Deserialize)]
        struct AuditResponse {
            #[serde(default)]
            findings: Vec<FindingResponse>,
        }
        #[derive(Deserialize)]
        struct FindingResponse {
            id: String,
            file: String,
            dimension: String,
            severity: String,
            description: String,
            #[serde(default)]
            suggested_fix: String,
        }

        let parsed: AuditResponse = validate_json(&response)?;

        Ok(AuditResult {
            findings: parsed.findings.into_iter().map(|f| AuditFinding {
                id: f.id,
                file: f.file,
                dimension: f.dimension,
                severity: match f.severity.as_str() {
                    "high" => Severity::High,
                    "medium" => Severity::Medium,
                    _ => Severity::Low,
                },
                description: f.description,
                suggested_fix: f.suggested_fix,
            }).collect(),
            files_audited: files.to_vec(),
            dimensions_checked: dimensions.to_vec(),
        })
    }

    fn fix(
        &mut self,
        finding_id: &str,
        finding_description: &str,
        file_path: &str,
        file_content: &str,
        _skill_context: &str,
    ) -> anyhow::Result<FixResult> {
        let system = "You are a code fixer. Apply the suggested fix. \
             Respond with JSON: {\"success\": true, \"files_modified\": [\"path\"], \"description\": \"what changed\"}";
        let user = format!(
            "Fix finding {finding_id}: {finding_description}\n\nFile: {file_path}\n```\n{}\n```",
            &file_content[..file_content.len().min(8000)]
        );

        let rt = tokio::runtime::Handle::try_current()
            .map_err(|_| anyhow::anyhow!("No tokio runtime"))?;
        let response = rt.block_on(self.call_api(system, &user))?;
        validate_json(&response)
    }

    fn evaluate_independence(
        &mut self,
        pass_a: &AuditResult,
        pass_b: &AuditResult,
    ) -> anyhow::Result<EvaluationResult> {
        let system = "You evaluate whether two audit passes are independently produced. \
             Respond with JSON: {\"independent\": true/false, \"rationale\": \"why\"}";
        let user = format!(
            "Pass A findings: {:?}\nPass B findings: {:?}\n\nAre these independently produced?",
            pass_a.findings.iter().map(|f| &f.description).collect::<Vec<_>>(),
            pass_b.findings.iter().map(|f| &f.description).collect::<Vec<_>>(),
        );

        let rt = tokio::runtime::Handle::try_current()
            .map_err(|_| anyhow::anyhow!("No tokio runtime"))?;
        let response = rt.block_on(self.call_api(system, &user))?;
        validate_json(&response)
    }

    fn write(&mut self, spec: &str, skill_context: &str) -> anyhow::Result<WriteResult> {
        let system = format!(
            "You write content based on specifications. \
             Respond with JSON: {{\"content\": \"the content\", \"files_written\": [\"paths\"], \"description\": \"what was written\"}}\n\
             {}",
            if skill_context.is_empty() { "" } else { skill_context },
        );

        let rt = tokio::runtime::Handle::try_current()
            .map_err(|_| anyhow::anyhow!("No tokio runtime"))?;
        let response = rt.block_on(self.call_api(&system, spec))?;
        validate_json(&response)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_model_tier_ids() {
        assert!(ModelTier::Fast.model_id().contains("haiku"));
        assert!(ModelTier::Standard.model_id().contains("sonnet"));
        assert!(ModelTier::Frontier.model_id().contains("opus"));
    }

    #[test]
    fn test_validate_json_clean() {
        #[derive(Deserialize)]
        struct Test { value: i32 }
        let result: Test = validate_json(r#"{"value": 42}"#).unwrap();
        assert_eq!(result.value, 42);
    }

    #[test]
    fn test_validate_json_wrapped_in_markdown() {
        #[derive(Deserialize)]
        struct Test { value: i32 }
        let result: Test = validate_json("```json\n{\"value\": 42}\n```").unwrap();
        assert_eq!(result.value, 42);
    }

    #[test]
    fn test_validate_json_invalid() {
        #[derive(Deserialize)]
        struct Test { value: i32 }
        assert!(validate_json::<Test>("not json").is_err());
    }

    #[test]
    fn test_validate_and_retry_succeeds() {
        let mut count = 0;
        let result: serde_json::Value = validate_and_retry(
            || {
                count += 1;
                Ok(r#"{"ok": true}"#.to_string())
            },
            2,
        ).unwrap();
        assert_eq!(count, 1);
        assert_eq!(result["ok"], true);
    }

    #[test]
    fn test_validate_and_retry_fails_then_succeeds() {
        let mut count = 0;
        let result: serde_json::Value = validate_and_retry(
            || {
                count += 1;
                if count == 1 {
                    Ok("not json".to_string())
                } else {
                    Ok(r#"{"ok": true}"#.to_string())
                }
            },
            2,
        ).unwrap();
        assert_eq!(count, 2);
    }

    #[test]
    fn test_validate_and_retry_exhausted() {
        let result: anyhow::Result<serde_json::Value> = validate_and_retry(
            || Ok("invalid".to_string()),
            2,
        );
        assert!(result.is_err());
        assert!(result.unwrap_err().to_string().contains("3 attempts"));
    }

    #[test]
    fn test_from_env_missing_key() {
        // Don't set ANTHROPIC_API_KEY
        unsafe { std::env::remove_var("ANTHROPIC_API_KEY"); }
        assert!(AnthropicProvider::from_env(ModelTier::Standard).is_err());
    }
}
